In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import os

In [3]:
headers = [
    "Kod stacji",
    "Nazwa stacji",
    "Rok",
    "Miesiąc",
    "Dzień",
    "Średnia dobowa temperatura [°C]",
    "Status pomiaru TEMP",
    "Średnia dobowa wilgotność względna [%]",
    "Status pomiaru WLGS",
    "Średnia dobowa prędkość wiatru [m/s]",
    "Status pomiaru FWS",
    "Średnie dobowe zachmurzenie ogólne [oktanty]",
    "Status pomiaru NOS"
]

Pogoda_Path = 'Dane/Pogoda/'
dataframes=[]

for filename in os.listdir(Pogoda_Path):
    file_path = os.path.join(Pogoda_Path, filename)
    df = pd.read_csv(file_path, encoding = 'cp1250', names = headers)
    dataframes.append(df)

dfp = pd.concat(dataframes, ignore_index = True)

In [4]:
dfp

Kod stacji Nazwa stacji   Rok  Miesiąc  Dzień  \
0         249180010     PSZCZYNA  2001        1      1   
1         249180010     PSZCZYNA  2001        1      2   
2         249180010     PSZCZYNA  2001        1      3   
3         249180010     PSZCZYNA  2001        1      4   
4         249180010     PSZCZYNA  2001        1      5   
...             ...          ...   ...      ...    ...   
1015638   254220090       OLECKO  2022       12     27   
1015639   254220090       OLECKO  2022       12     28   
1015640   254220090       OLECKO  2022       12     29   
1015641   254220090       OLECKO  2022       12     30   
1015642   254220090       OLECKO  2022       12     31   

         Średnia dobowa temperatura [°C]  Status pomiaru TEMP  \
0                                   -5.7                  NaN   
1                                   -2.7                  NaN   
2                                   -1.5                  NaN   
3                                    1.9                  NaN   
4                                    2.8                  NaN   
...                                  ...                  ...   
1015638                              2.6                  NaN   
1015639                              0.7                  NaN   
1015640                              3.7                  NaN   
1015641                              4.3                  NaN   
1015642                              5.6                  NaN   

         Średnia dobowa wilgotność względna [%]  Status pomiaru WLGS  \
0                                           0.0                  8.0   
1                                           0.0                  8.0   
2                                           0.0                  8.0   
3                                           0.0                  8.0   
4                                           0.0                  8.0   
...                                         ...                  ...   
1015638                                     0.0                  8.0   
1015639                                     0.0                  8.0   
1015640                                     0.0                  8.0   
1015641                                     0.0                  8.0   
1015642                                     0.0                  8.0   

         Średnia dobowa prędkość wiatru [m/s]  Status pomiaru FWS  \
0                                         1.0                 NaN   
1                                         2.0                 NaN   
2                                         1.7                 NaN   
3                                         1.0                 NaN   
4                                         1.0                 NaN   
...                                       ...                 ...   
1015638                                   0.0                 8.0   
1015639                                   0.0                 8.0   
1015640                                   0.0                 8.0   
1015641                                   0.0                 8.0   
1015642                                   0.0                 8.0   

         Średnie dobowe zachmurzenie ogólne [oktanty]  Status pomiaru NOS  
0                                                 2.0                 NaN  
1                                                 7.0                 NaN  
2                                                 8.0                 NaN  
3                                                 3.7                 NaN  
4                                                 3.7                 NaN  
...                                               ...                 ...  
1015638                                           8.0                 NaN  
1015639                                           8.0                 NaN  
1015640                                           8.0                 NaN  
1015641                                           7.0                 NaN  
1

In [5]:
dfp['Data'] = dfp[dfp.columns[2:5]].apply(lambda x: '-'.join(x.dropna().astype(str)),axis=1)
dfp.drop(columns=['Nazwa stacji','Kod stacji','Rok','Miesiąc','Dzień','Status pomiaru NOS','Średnie dobowe zachmurzenie ogólne [oktanty]','Status pomiaru FWS','Średnia dobowa prędkość wiatru [m/s]','Status pomiaru WLGS','Status pomiaru TEMP'], inplace = True)

In [6]:
dfp['Data'] = dfp['Data'].apply(lambda x: '-'.join([part.zfill(2) for part in x.split('-')]))


In [7]:
dfp['Data'] = pd.to_datetime(dfp['Data'], format='%Y-%m-%d')

In [8]:
dfp

Średnia dobowa temperatura [°C]  \
0                                   -5.7   
1                                   -2.7   
2                                   -1.5   
3                                    1.9   
4                                    2.8   
...                                  ...   
1015638                              2.6   
1015639                              0.7   
1015640                              3.7   
1015641                              4.3   
1015642                              5.6   

         Średnia dobowa wilgotność względna [%]       Data  
0                                           0.0 2001-01-01  
1                                           0.0 2001-01-02  
2                                           0.0 2001-01-03  
3                                           0.0 2001-01-04  
4                                           0.0 2001-01-05  
...                                         ...        ...  
1015638                                     0.0 2022-12-27  
1015639                                     0.0 2022-12-28  
1015640                                     0.0 2022-12-29  
1015641                                     0.0 2022-12-30  
1015642                                     0.0 2022-12-31  

[1015643 rows x 3 columns]

In [9]:
# dfp.drop(dfp.columns[2], axis = 1, inplace = True)?

In [10]:
dfp=dfp[dfp['Średnia dobowa wilgotność względna [%]'] != 0.0]

In [11]:
dfp

Średnia dobowa temperatura [°C]  \
31                                  -5.0   
32                                   0.2   
33                                   0.7   
34                                   3.6   
35                                   4.0   
...                                  ...   
1015607                              3.8   
1015608                              2.0   
1015609                              5.8   
1015610                              6.0   
1015611                              8.4   

         Średnia dobowa wilgotność względna [%]       Data  
31                                         83.0 2001-01-01  
32                                         79.0 2001-01-02  
33                                         94.5 2001-01-03  
34                                         84.5 2001-01-04  
35                                         86.8 2001-01-05  
...                                         ...        ...  
1015607                                    86.8 2022-12-27  
1015608                                    90.0 2022-12-28  
1015609                                    87.5 2022-12-29  
1015610                                    76.8 2022-12-30  
1015611                                    94.0 2022-12-31  

[410734 rows x 3 columns]

In [12]:
dfp.isnull().sum()


Średnia dobowa temperatura [°C]           0
Średnia dobowa wilgotność względna [%]    0
Data                                      0
dtype: int64

In [13]:
#POZARY

In [14]:


Pozary_Path = 'Dane/Pozary/'
dataframes=[]

for filename in os.listdir(Pozary_Path):
    file_path = os.path.join(Pozary_Path, filename)
    dff = pd.read_csv(file_path, encoding = 'cp1250')
    dataframes.append(dff)

dff = pd.concat(dataframes, ignore_index = True)

In [15]:
dff

latitude  longitude  brightness  scan  track    acq_date  acq_time  \
0       51.6878    15.9832       320.9   1.4    1.2  2001-01-11       955   
1       51.6847    15.9881       308.3   1.6    1.3  2001-01-20       949   
2       51.6867    15.9813       310.3   1.6    1.3  2001-01-20       949   
3       53.1366    22.8249       324.2   1.6    1.2  2001-01-24       924   
4       53.1346    22.8485       319.8   1.6    1.2  2001-01-24       924   
...         ...        ...         ...   ...    ...         ...       ...   
22559   51.6870    15.9799       307.6   1.1    1.1  2022-11-02      1005   
22560   51.6881    15.9838       300.9   1.3    1.1  2022-11-15       141   
22561   51.1780    16.1117       314.5   1.1    1.0  2022-11-25      1147   
22562   51.6873    15.9845       301.5   1.0    1.0  2022-12-29       955   
22563   51.1734    16.1039       300.8   1.0    1.0  2022-12-30      1204   

      satellite instrument  confidence  version  bright_t31   frp daynight  \
0         Terra      MODIS          76     6.20       273.6  39.1        D   
1         Terra      MODIS          55     6.20       268.6  29.5        D   
2         Terra      MODIS          66     6.20       269.0  30.0        D   
3         Terra      MODIS          83     6.20       270.6  54.4        D   
4         Terra      MODIS          80     6.20       270.0  45.2        D   
...         ...        ...         ...      ...         ...   ...      ...   
22559     Terra      MODIS          57     6.03       287.0  11.4        D   
22560      Aqua      MODIS          35     6.03       274.0  12.9        N   
22561      Aqua      MODIS          75     6.03       274.8  17.5        D   
22562     Terra      MODIS          47     6.03       278.8   8.5        D   
22563      Aqua      MODIS          42     6.03       280.5   7.8        D   

       type  
0         2  
1         2  
2         2  
3         0  
4         0  
...     ...  
22559     2  
22560     2  
22561     2  
22562     2  
22563     2  

[22564 rows x 15 columns]

In [16]:
dff.drop(columns = ['brightness','scan','track','acq_time','satellite','instrument','confidence','version','bright_t31','frp','daynight','latitude','longitude'], inplace = True)

In [17]:
dff=dff[dff['type'] == 0]

In [19]:
dff

acq_date  type
3      2001-01-24     0
4      2001-01-24     0
8      2001-02-26     0
9      2001-03-01     0
10     2001-03-01     0
...           ...   ...
22538  2022-09-07     0
22539  2022-09-07     0
22544  2022-09-21     0
22545  2022-09-21     0
22554  2022-10-26     0

[16955 rows x 2 columns]

In [20]:
dff.rename(columns={'acq_date':'Data','type':'Typ'}, inplace = True)

C:\Users\OKej\AppData\Local\Temp\ipykernel_35068\1194424143.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dff.rename(columns={'acq_date':'Data','type':'Typ'}, inplace = True)


In [21]:
dff.isnull().sum()

Data    0
Typ     0
dtype: int64

In [28]:
dff_fire_reduced = dff.groupby('Data', as_index=False).size()


In [29]:
dff_fire_reduced

Data  size
0     2001-01-24     2
1     2001-02-26     1
2     2001-03-01     3
3     2001-03-12     3
4     2001-03-25     2
...          ...   ...
2842  2022-09-05     2
2843  2022-09-06     1
2844  2022-09-07     5
2845  2022-09-21     2
2846  2022-10-26     1

[2847 rows x 2 columns]

In [30]:
dff_fire_reduced['Data'] = pd.to_datetime(dff_fire_reduced['Data'], format='%Y-%m-%d')

In [26]:
dff_fire_reduced

Data  size
0    2001-01-24     2
1    2001-02-26     1
2    2001-03-01     3
3    2001-03-12     3
4    2001-03-25     2
...         ...   ...
2842 2022-09-05     2
2843 2022-09-06     1
2844 2022-09-07     5
2845 2022-09-21     2
2846 2022-10-26     1

[2847 rows x 2 columns]

In [33]:

df_combined = pd.merge(dfp, dff_fire_reduced, on='Data', how='left')


In [34]:
df_combined.rename(columns={'size': 'Pozar'}, inplace=True)

In [35]:
df_combined.fillna(0, inplace = True)

In [37]:
df_combined.to_csv('test_df_combined.csv')

In [52]:
from sklearn.model_selection import train_test_split

X = df_combined[['Średnia dobowa temperatura [°C]', 'Średnia dobowa wilgotność względna [%]']]
y = df_combined['Pozar']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)


In [53]:
df_combined


Średnia dobowa temperatura [°C]  \
0                              0.392796   
1                              0.481990   
2                              0.490566   
3                              0.540309   
4                              0.547170   
...                                 ...   
410729                         0.543739   
410730                         0.512864   
410731                         0.578045   
410732                         0.581475   
410733                         0.622642   

        Średnia dobowa wilgotność względna [%]       Data  Pozar  
0                                     0.824199 2001-01-01      0  
1                                     0.782834 2001-01-02      0  
2                                     0.943123 2001-01-03      0  
3                                     0.839710 2001-01-04      0  
4                                     0.863495 2001-01-05      0  
...                                        ...        ...    ...  
410729                                0.863495 2022-12-27      0  
410730                                0.896587 2022-12-28      0  
410731                                0.870734 2022-12-29      0  
410732                                0.760083 2022-12-30      0  
410733                                0.937952 2022-12-31      0  

[410734 rows x 4 columns]

In [54]:
df_combined['Pozar'] = np.where(df_combined['Pozar'] > 0, 1, 0)

In [55]:
print(df_combined['Średnia dobowa temperatura [°C]'].min())
print(df_combined['Średnia dobowa temperatura [°C]'].max())


0.0
1.0


In [56]:
from sklearn.preprocessing import MinMaxScaler, StandardScaler

# Skaler dla normalizacji
scaler = MinMaxScaler()


# Wyodrębnienie kolumn
columns_to_scale = df_combined[['Średnia dobowa temperatura [°C]', 'Średnia dobowa wilgotność względna [%]']]


# Normalizacja
normalized_data = scaler.fit_transform(columns_to_scale)


# Zamiana danych w DataFrame na znormalizowane
df_combined[['Średnia dobowa temperatura [°C]', 'Średnia dobowa wilgotność względna [%]']] = normalized_data


df_combined.to_csv('test.csv')
print("ok")

ok


1. **MLPRegressor (Multi-layer Perceptron Regressor):**
   - Jest to model regresji oparty na sztucznych sieciach neuronowych, składający się z co najmniej trzech warstw: warstwy wejściowej, ukrytej (lub warstw) oraz warstwy wyjściowej. 
   - Model ten może być stosowany do problemów regresji, gdzie dąży się do przewidywania ciągłych wartości wyjściowych.

2. **RandomForestRegressor:**
   - Jest to model regresji oparty na algorytmie lasów losowych, który składa się z wielu drzew decyzyjnych. 
   - Każde drzewo w lesie jest trenowane na różnych podzbiorach danych, co prowadzi do zrównoważonej i silnej generalizacji. 
   - Model ten jest elastyczny i nadaje się do wielu rodzajów danych, przy zachowaniu odporności na overfitting.

3. **DecisionTreeRegressor:**
   - Jest to model regresji oparty na drzewach decyzyjnych, który dzieli zbiór danych na podzbiory, stosując serię reguł decyzyjnych opartych na wartościach cech. 
   - Model ten buduje drzewo decyzyjne, które jest używane do prognozowania wartości ciągłych na podstawie danych wejściowych. 
   - Może być podatny na overfitting, ale może być regulowany przez parametryzację.

4. **LinearRegression:**
   - Jest to podstawowy model regresji liniowej, który zakłada liniową zależność między zmiennymi niezależnymi a zmienną zależną. 
   - Model ten stara się dopasować linię najlepszego dopasowania do danych poprzez minimalizację sumy kwadratów różnic między wartościami przewidywanymi a rzeczywistymi. 
   - Jest łatwy w interpretacji i implementacji, ale może być ograniczony, gdy dane nie spełniają założeń liniowości.


In [57]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score
from sklearn.tree import DecisionTreeRegressor
from sklearn.metrics import mean_squared_error
from sklearn.ensemble import RandomForestRegressor
from sklearn.neural_network import MLPRegressor
from sklearn.metrics import mean_absolute_error
from sklearn.svm import SVR

In [58]:
# REGRESJA LINIOWA

model = LinearRegression()

model.fit(X_train, y_train)

y_pred = model.predict(X_test)

mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print('Błąd średniokwadratowy:', mse)

print('Współczynniki regresji:', model.coef_)
print('Wyraz wolny:', model.intercept_)
print('Współczynnik determinacji (R-squared):', r2)


Błąd średniokwadratowy: 0.18133319135386433
Współczynniki regresji: [ 1.09563161 -0.93694572]
Wyraz wolny: 0.4417269873976136
Współczynnik determinacji (R-squared): 0.22627235981689864


In [59]:
# DRZEWA DECYZYJNE

model = DecisionTreeRegressor()

model.fit(X_train, y_train)

y_pred = model.predict(X_test)

mse = mean_squared_error(y_test, y_pred)
print('Błąd średniokwadratowy:', mse)

r2 = r2_score(y_test, y_pred)
print('Współczynnik determinacji (R-squared):', r2)

 
mae = mean_absolute_error(y_test, y_pred)
print('Średni błąd bezwzględny (MAE):', mae)

rmse = np.sqrt(mse)
print('Średni błąd kwadratowy pierwiastkowany (RMSE):', rmse)



Błąd średniokwadratowy: 0.2257585617224647
Współczynnik determinacji (R-squared): 0.03671447070171929
Średni błąd bezwzględny (MAE): 0.3578090734441572
Średni błąd kwadratowy pierwiastkowany (RMSE): 0.47514057048673997


In [60]:
# LAS LOSOWY

model = RandomForestRegressor()

model.fit(X_train, y_train)

y_pred = model.predict(X_test)

mse = mean_squared_error(y_test, y_pred)
print('Błąd średniokwadratowy:', mse)

r2 = r2_score(y_test, y_pred)
print('Współczynnik determinacji (R-squared):', r2)

 
mae = mean_absolute_error(y_test, y_pred)
print('Średni błąd bezwzględny (MAE):', mae)

rmse = np.sqrt(mse)
print('Średni błąd kwadratowy pierwiastkowany (RMSE):', rmse)



Błąd średniokwadratowy: 0.2126788693779828
Współczynnik determinacji (R-squared): 0.09252399689192425
Średni błąd bezwzględny (MAE): 0.358199986539732
Średni błąd kwadratowy pierwiastkowany (RMSE): 0.46117119313545896


In [62]:
# SIECI NEURONOWE

model = MLPRegressor(hidden_layer_sizes=(100, 50,20), max_iter=500, alpha=0.001, early_stopping=True)
model.fit(X_train, y_train)

y_pred = model.predict(X_test)

mse = mean_squared_error(y_test, y_pred)
print('Błąd średniokwadratowy:', mse)

r2 = r2_score(y_test, y_pred)
print('Współczynnik determinacji (R-squared):', r2)

 
mae = mean_absolute_error(y_test, y_pred)
print('Średni błąd bezwzględny (MAE):', mae)

rmse = np.sqrt(mse)
print('Średni błąd kwadratowy pierwiastkowany (RMSE):', rmse)

print('Strata po ostatniej epoce:', model.loss_)



Błąd średniokwadratowy: 0.17903583145690596
Współczynnik determinacji (R-squared): 0.2360749273361351
Średni błąd bezwzględny (MAE): 0.3581737823297417
Średni błąd kwadratowy pierwiastkowany (RMSE): 0.4231262594745284
Strata po ostatniej epoce: 0.08995720082688077


## Wyniki analizy modelu 2010-2022

- **Błąd średniokwadratowy (MSE):** 0.1798
- **Współczynnik determinacji (R-squared):** 0.2010
- **Średni błąd bezwzględny (MAE):** 0.3622
- **Średni błąd kwadratowy pierwiastkowany (RMSE):** 0.4241
- **Strata po ostatniej epoce:** 0.0907

## Wyniki analizy modelu 2001-2022

- **Błąd średniokwadratowy (MSE):** 0.1790
- **Współczynnik determinacji (R-squared):** 0.2346
- **Średni błąd bezwzględny (MAE):** 0.3582
- **Średni błąd kwadratowy pierwiastkowany (RMSE):** 0.4231
- **Strata po ostatniej epoce:** 0.0810

1. **Błąd średniokwadratowy (MSE):** Średni kwadrat różnicy między przewidywanymi a rzeczywistymi wartościami.

2. **Współczynnik determinacji (R-squared):** Miara, która określa, jak dobrze model pasuje do danych, porównując wariancję przewidywań modelu do wariancji rzeczywistych danych.

3. **Średni błąd bezwzględny (MAE):** Średnia wartość bezwzględna różnicy między przewidywanymi a rzeczywistymi wartościami.

4. **Średni błąd kwadratowy pierwiastkowany (RMSE):** Pierwiastek kwadratowy z MSE, co daje miarę przeciętnego błędu przewidywania w tych samych jednostkach, co oryginalne dane.

5. **Strata po ostatniej epoce:** Wartość błędu oszacowana przez model podczas ostatniej epoki treningowej.




## IMPLEMENTACJA MODELU

### 1. Przesyłanie danych do API
System pobiera dane dotyczące aktualnej temperatury i wilgotności poprzez api NASA oraz api GIOS

### 2. Przetwarzanie przez model
API Django odbiera dane, przetwarza je (np. konwertuje na odpowiedni format) i używa wcześniej zapisanego modelu do wykonania predykcji. Model ocenia, czy dane warunki mogą prowadzić do pożaru.

### 3. Zwracanie wyniku
API zwraca wynik predykcji do interfejsu użytkownika React, który może to być na przykład wartość binarna (0 lub 1, gdzie 1 oznacza w ryzyko pożaru).

### 4. Wyświetlanie wyniku użytkownikowi
Aplikacja React odbiera odpowiedź z API i wyświetla użytkownikowi wynik, informując, czy istnieje ryzyko pożaru.
s

In [63]:
# IMPORT MODELU

import joblib

joblib.dump(model, 'model.pkl')


# POBIERANIE DANYCH - PRZYKŁAD 

#PLIK DATA_FETCHER

import requests

def get_nasa_fires_data():
    url = "https://nasa.gov/fires_api/latest_fires"
    response = requests.get(url)
    return response.json() if response.status_code == 200 else None

def get_weather_data(location):
    api_key = "weatherapi_key"
    url = f"http://api.weatherapi.com/v1/current.json?key={api_key}&q={location}"
    response = requests.get(url)
    return response.json() if response.status_code == 200 else None

def get_gios_air_quality_data(station_id):
    url = f"http://api.gios.gov.pl/pjp-api/rest/aqindex/getIndex/{station_id}"
    response = requests.get(url)
    return response.json() if response.status_code == 200 else None


# ZAŁADOWANIE MODELU | DEFINICJA FUNKCJI

model = joblib.load('path/model.pkl')

def predict_fire_risk(temperature, humidity):
    input_data = np.array([[temperature, humidity]])
    prediction = model.predict(input_data)
    return prediction[0]


# UZYCIE FUNKCJI W WIDOKU API DJANGO

from django.http import JsonResponse
from rest_framework.views import APIView
from .data_fetcher import get_nasa_fires_data, get_weather_data, get_gios_air_quality_data
from .model import predict_fire_risk, prepare_data

class FireRiskPredictionAPIView(APIView):
    def get(self, request, *args, **kwargs):
        fire_data = get_nasa_fires_data()
        weather_data = get_weather_data('Location')

        if fire_data and weather_data
            input_data = prepare_data(fire_data, weather_data)
            prediction = predict_fire_risk(input_data)
            return JsonResponse({'prediction': prediction})
        else:
            return JsonResponse({'error': 'Error fetching data'}, status=400)

        


SyntaxError: expected ':' (1655282596.py, line 48)